### ISSUE TITLE
Interface Options for Bibliography

### ISSUE THREAD

**Issue by FelixBenning** [FelixBenning/bibcite#1]
collect different appraoches

**Comment by FelixBenning** [comment]
# 1 Web Based

- Bibliography files are a reference e.g. `references.bib` to be fetched
  1. either
    ```html
     <script src="references.bib" </script>
     <!-- or -->
     <script> /* references */ </script>
    ``` 
   2.  or custom 
   ```html
    <bib-config bib="references.bib" citation-style="author-year" ... > </bib-config>
    <!-- or -->
    <bib-config citation-style="author-year" ... > <!-- references --> </bib-config>
    ```

   I currently think the second approach is probably better

## Advantages
- Clear Analogue to CSS and JS
-  Possible to drop-in this library using `<script>` tags and it just works

## Disadvantage
- Performance `reference.bib` needs to be
    - fetched (advantage: browsers will cache it)
    - parsed
    - pruned (which keys are actually in use?)
- Completely broken with script blockers

**Comment by FelixBenning** [comment]
# 2a Bundler Based

Install package with npm and do something like

```javascript
import bibcite from 'bibcite'
import bibliography from 'references.bib'

document.bib = bibcite(bib=bibliography, citationStyle='author-year', ...)
```

This already avoids the `fetch` as the bibliography would be shipped together with the javascript. 

The web based usage would become more awkward though

```html
<script src="bibcite.js"></script>
<script>
bibliography = fetch('references.bib')
document.bib = bibcite(bib=bibliography, citationStyle='author-year',...)
</script>
```

## Advantages

- saves one communication roundtrip with the server

## Disadvantages

- Usability
- Still broken with script blockers

**Comment by FelixBenning** [comment]
# 2b Bundler Based with Custom Loader

Using something like a `webpack custom loader` one might be able to pre-parse the bibliography on import from bibcite. 

```javascript
bibcite(bib=bibliography,...)
```

would in this case expect a parsed `bibliography` not a string.



## Advantages

- No fetching, no parsing

## Disadvantages

- Even less usability (bundler dependent configuration hassle)
- Maintenance (bundler dependent)
- Still broken with script blockers since the parsed result is not part of the html file but the javascript file

**Comment by FelixBenning** [comment]
# 3 Do everything server side

Using something like [libxmljs](https://github.com/libxmljs/libxmljs) replace all citations and generate bibliography

## Advantages

- Performance
- Works with script blockers

## Disadvantages

- Usability: Is a build step (either one more build step, or a build step for a project which previously required no build)
- No interactivity possible (previews in online editors)

**Comment by FelixBenning** [comment]
MathJax does a combination of 1 and 3, most people use method 1 with mathjax. Which makes me think that everything else is probably premature optimization. So focusing on that now.

**Comment by FelixBenning** [comment]
Using a templating engine you can probably move the content of `references.bib` into the html file. This makes Approach 1 equivalent to approach 2 performance wise. And templating engines are included in static site generators so they are probably the more natural tool for people doing citations than javascript bundlers
